rfc/src/rfcs/0008-function-contracts.md

```diff
@@ -0,0 +1,580 @@
+
+   A possible harness that uses our `my_div` contract could be the following:
+
+   ```rs
+   #[kani::proof]
+   #[kani::stub_verified(my_div)]
+   fn use_div() {
+     let v = vec![...];
+     let some_idx = my_div(v.len() - 1, 3);
+     v[some_idx];
+   }
+   ```
+
+   To use the contract as a stub Kani must first ensure the calling context is
+   safe. It inserts a `kani::assert` for the preconditions (`requires`) before
+   the call; then it replaces the result of `my_div` with a non-deterministic
+   value. Finally the non-deterministic result is constrained by a
+   `kani::assume` of the postconditions (`ensures`).
+   
+   Mutable memory is similarly made non-deterministic, discussed later in
+   [havociing](#memory-predicates-and-havocing).
+
+   The expanded version is equivalent to this:
+  
+   ```rs
+   #[kani::proof]
+   #[kani::stub_verified(my_div)]
+   fn use_div() {
+     let v = vec![...];
+     let dividend = v.len() - 1;
+     let divisor = 3;
+     kani::assert(divisor != 0);
+     let result = kani::any();
+     kani::assume(result <= dividend);
+     let some_idx = result;
+     v[some_idx];
+   }
+   ```
+
+   Notice that this performs no actual computation for `my_div` (other than the
+   conditions) which allows us to avoid something potentially costly.
+
+Also notice that Kani was able to express both contract checking and stubbing
+with existing capabilities; the important feature is the enforcement. The
+checking is, by construction, performed **against the same condition** that is
+later used as stub, which ensures soundness (see discussion on lingering threats
+to soundness in the [future](#future-possibilities) section) and guarding against
+stubs diverging from their checks.
+
+### History Variables
+
+Kani's contract language contains additional support for reasoning about changes
+to memory. One case where this is necessary is whenever `ensures` needs to
+reason about state before the function call. By default it only has access to
+state after the call completes, which will be different if the call mutates
+memory.
+
+Consider the `Vec::pop` function
+
+```rs
+impl<T> Vec<T> {
+  fn pop(&mut self) -> Option<T> {
+    ...
+  }
+}
+```
+
+If we want to describe in which case the result is `Some`, we need to know
+whether `self` is empty *before* `pop` is called. To do this Kani provides the
+`old(EXPR)` pseudo function, which evaluates `EXPR` before the call (e.g. to
+`pop`) and makes the result available to `ensures`. It is used like so:
+
+```rs
+impl<T> Vec<T> {
+  #[kani::ensures(old(self.is_empty()) || result.is_some())]
+  fn pop(&mut self) -> Option<T> {
+    ...
+  }
+}
+```
+
+`old` allows evaluating any (side-effect free[^side-effects]) Rust expression.
+The bowwor checker enforces the result of `old` cannot observe the mutations
+from e.g. `pop`, as that would defeat the purpose. If `your` expression in `old`
+returns borrowed content, make a copy instead (using e.g. `clone()`).
```

Comment:
I'm confused by this explanation... I'm guessing that you are trying to say is that `old(self).is_empty()` is not equivalent to `old(self.is_empty())`.

Since `self` is a reference, `old(self)` will only refer to the value of the reference, not the value of the object pointed to.